processing/src/main/java/org/apache/druid/query/lookup/ImmutableLookupMap.java

```diff
@@ -0,0 +1,370 @@
+    {
+      return true;
+    }
+
+    @Override
+    public boolean canGetKeySet()
+    {
+      return true;
+    }
+
+    @Override
+    public Iterable<Map.Entry<String, String>> iterable()
+    {
+      return () -> IntStream.range(0, keys.size())
+                            .<Map.Entry<String, String>>mapToObj(i -> new Entry(keys.get(i), values.get(i)))
+                            .iterator();
+    }
+
+    @Override
+    public Set<String> keySet()
+    {
+      return keyToEntry.keySet();
+    }
+
+    @Override
+    public boolean isOneToOne()
+    {
+      return isOneToOne;
+    }
+
+    @Override
+    public long estimateHeapFootprint()
+    {
+      return MapLookupExtractor.estimateHeapFootprint(iterable());
+    }
+
+    @Override
+    public byte[] getCacheKey()
+    {
+      return cacheKeySupplier.get();
+    }
+
+    public Map<String, String> asMap()
+    {
+      return ImmutableLookupMap.this.asMap;
+    }
+  }
+
+  /**
+   * Builder. Once done populating, call {@link #build()} to get a {@link LookupExtractor}.
+   */
+  public static class Builder
+  {
+    private Map<String, String> entries;
+    private boolean built = false;
+
+    /**
+     * Create a builder.
+     */
+    public Builder()
+    {
+      this.entries = new HashMap<>();
+    }
+
+    /**
+     * Create a builder starting from a particular map. The provided map will be modified if calls are made to
+     * {@link #put(String, String)} or {@link #putAll(Map)}.
+     */
+    public Builder(final Map<String, String> map)
+    {
+      this.entries = map;
+    }
+
+    public Builder put(final String key, final String value)
+    {
+      entries.put(key, value);
+      return this;
+    }
+
+    public Builder putAll(final Map<String, String> map)
+    {
+      entries.putAll(map);
+      return this;
+    }
+
+    /**
+     * Build the extractor. The returned extractor will not share the map that was passed to the constructor
+     * {@link #Builder(Map)}, if any.
+     */
+    public ImmutableLookupMap build()
+    {
+      if (built) {
+        throw new ISE("Already built");
+      } else {
+        built = true;
+      }
+
+      final List<Map.Entry<String, String>> entriesList = new ArrayList<>(entries.size());
+      entriesList.addAll(entries.entrySet());
+      entriesList.sort(VALUE_COMPARATOR);
+
+      entries = null; // save memory
+
+      final List<String> keys = new ArrayList<>();
+      final List<String> values = new ArrayList<>();
+
+      for (final Map.Entry<String, String> entry : entriesList) {
+        keys.add(entry.getKey());
+        values.add(entry.getValue());
+      }
+
+      entriesList.clear(); // save memory
+
+      // Populate keyToEntries map.
+      final Object2IntMap<String> keyToEntry = new Object2IntOpenHashMap<>(keys.size(), LOAD_FACTOR);
+      keyToEntry.defaultReturnValue(NOT_FOUND);
+      for (int i = 0; i < keys.size(); i++) {
+        keyToEntry.put(keys.get(i), i);
+      }
```

Comment:
any reason not to do this while looping to build `keys` and `values`? worried about memory during loading perhaps?

Reply:
Yeah, mainly a memory thing, so we don't have `entriesList` and `keyToEntry` at the same time.